pkg/utils/kubernetes/utils.go

```diff
@@ -0,0 +1,130 @@
+package kubernetes
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"strings"
+)
+
+//OnFinalize Callback function to be executed when an object is being deleted
+type OnFinalize func() error
+
+type ObjectFinalizer struct {
+	object     runtime.Object
+	finalizers map[string]OnFinalize
+}
+
+type FinalizerManager struct {
+	objects map[types.UID]ObjectFinalizer
+	Client  client.Client
+}
+
+func NewFinalizerManager(client client.Client) FinalizerManager {
+	return FinalizerManager{
+		objects: map[types.UID]ObjectFinalizer{},
+		Client:  client,
+	}
+}
+
+//RegisterFinalizer registers a finalizer function to be executed when the object is marked to be deleted.
+func (mgr *FinalizerManager) RegisterFinalizer(owner metav1.Object, name string, onFinalize OnFinalize) error {
+	if mgr.IsFinalizing(owner) {
+		return nil
+	}
+	err := validateFinalizer(name, onFinalize)
+	if err != nil {
+		return err
+	}
+	controllerutil.AddFinalizer(owner, name)
+	obj, err := toRuntimeObj(owner)
+	if err != nil {
+		return err
+	}
+	err = mgr.Client.Update(context.TODO(), obj)
+	if err != nil {
+		return err
+	}
+	_, ok := mgr.objects[owner.GetUID()]
+	if !ok {
+		mgr.objects[owner.GetUID()] = ObjectFinalizer{
+			object:     obj,
+			finalizers: map[string]OnFinalize{},
+		}
+	}
+	mgr.objects[owner.GetUID()].finalizers[name] = onFinalize
+	return nil
+}
+
+//UnregisterFinalizer removes a finalizer and updates the owner object
+func (mgr *FinalizerManager) UnregisterFinalizer(owner metav1.Object, name string) error {
+	err := validateFinalizerName(name)
+	if err != nil {
+		return err
+	}
+	obj, err := meta.Accessor(owner)
+	controllerutil.RemoveFinalizer(obj, name)
+	objFinalizer := mgr.objects[obj.GetUID()]
+	err = mgr.Client.Update(context.TODO(), objFinalizer.object)
+	if err != nil {
+		return err
+	}
+	delete(mgr.objects[obj.GetUID()].finalizers, name)
+	return nil
+}
+
+//Finalize triggers all the finalizers registered for the given object
+func (mgr *FinalizerManager) Finalize(owner metav1.Object) error {
```

Comment:
Mostly thinking out loud here, but does it make sense to call this FinalizeOnDelete() and have clients call it indiscriminately on every reconcile, when they have a finalizer set?

Reply:
This is the purpose of this method and how I am using it in the kie-cloud-operator. I can rename it if you feel it is clearer for consumers.